### ISSUE TITLE
Userid / Userstorage working?

### ISSUE THREAD

**Issue by basiteaal** [mishushakov/dialogflow-web-v2#12]
Hi,

Thanks for the great work.
I was wondering if there's any possibility to access the user's ID or Userstorage by this framework?

Thanks in advance,

Bas

**Comment by mishushakov** [comment]
What's Userid and Userstorage?

**Comment by basiteaal** [comment]
Hi!

Some documentation from the Google Actions Docs:
UserStorage: https://developers.google.com/actions/reference/rest/Shared.Types/AppResponse#FIELDS.user_storage

https://developers.google.com/actions/assistant/save-data

Let's say we have a back-end handling a webhook where we would like to save information of the user (not about the conversation); so that the next time the same user opens the voice application he/she can be identified by the application's webhook.
In other words you can compare the userstorage of Google Actions to some kind of cookie which is being stored on the i.e. Nest Home or Smartphone with Google Assistant.

**Comment by mishushakov** [comment]
Ok, now I understand. This repository does not simulate Actions on Google and only works with Dialogflow Gateway, which utilises Dialogflow V2 API with some crucial helpers on top.

Currently the only way to identify a user can be done through looking the session ID and assigning it to his identity. Session IDs are persistent and saved in local storage the first time user opens Dialogflow for Web and when history mode and cookies are enabled.

The flow therefor can be like this:

1. User opens your Chatbot for the first time
2. Your chatbot asks for his name, e-mail etc.
3. You save his session id and his personal information with web hook in your database, like this:

```json
{
  "mikhails_session_id": {
      "name": "Mikhail"
  }
}
```

4. User opens your Chatbot and asks to do a restricted function, only signed-in users have access to
5. You webhook checks the Session ID of the user and if it exists, does the function on user's behalf

Same can be used for login.
Hope your issue is now resolved. Have a good day.

**Comment by basiteaal** [comment]
Great answer @mishushakov ! Thanks.

**Comment by mishushakov** [comment]
No worries. You have to understand, that Dialogflow isn't sometimes as capable as a Google Assistant. The way it's executed on Google Home or Google Assistant-enabled device isn't really like it's executed on Dialogflow, which is more consumer-oriented platform, whereas Actions on Google is a fully featured SDK to build Google Assistant Actions.

**Comment by basiteaal** [comment]
Indeed; but the job here actually 'simulates' the same experience; which is really nice! They should have a look at this from Mountain View ;)

**Comment by mishushakov** [comment]
@basiteaal a dark side of me is telling, that we need to bring Google Assistant to the Web. Unfortunately it may only remain as a dream for many of us. Google Assistant in Chrome would be super-useful and with this project we are a whole step closer to make it possible. But it's also a very risky project to do, since Google can introduce their solution at any point of time and we will be wiped off completely.

The dudes from Mountain View are perfectly aware of we're doing here, but seem to have a lack of time, personnel and focus to do it as well.

Thanks for using the product btw, any feedback is appreciated.

**Comment by basiteaal** [comment]
Hehe, so true.

PS. Assistant comes available in Chrome 72 though: https://chromeunboxed.com/google-assistant-chromebook-72-how-to/

But still many users won't use this. Besides; many companies want their assistant experience on their website as well. I will be testing this application in the upcoming days/weeks and let you know about any feedback.

Cheers mate!
